support-modules/acquisition-events/src/main/scala/com/gu/support/acquisitions/utils/Retry.scala

```diff
@@ -0,0 +1,17 @@
+package com.gu.support.acquisitions.utils
+
+import scala.concurrent.{Future, ExecutionContext}
+import cats.data.EitherT
+import cats.instances.future._
+
+object Retry {
+  def apply[A, B](max: Int)(f: => EitherT[Future, A, B])(implicit ec: ExecutionContext): EitherT[Future, A, B] = {
+    def attempt(triesLeft: Int): EitherT[Future, A, B] = {
+      f.leftFlatMap(error => {
+        if (triesLeft > 0) attempt(triesLeft - 1) else EitherT.fromEither(Left(error))
```

Comment:
would it be possible to (also/instead?) return the first error, or maybe a list of the errors, so that if the first attempt locks us out, we can see what went wrong that time?  Sometimes the first can time out, then the later ones can fail with "duplicate record", so you need to know both to be sure.

Reply:
you might be able to do some kind of lazy list to avoid complex recursion, along these lines
```scala
val tries = Stream.from(1 to 3).map(f).takeWhile(_.isLeft).toList
tries.find(_.isRight).getOrElse(Left(tries.map(_.left.get)))
```

Reply:
Yeah I think that would be a good idea. I had a quick go at your suggested code, the only issue is with `takeWhile(_.isLeft)` - as we're using an `EitherT` with a `Future`, the type of `_.isLeft` is `Future[Boolean]`, I wasn't quite sure how to get around that?

Just for comparison, I had a go at the recursive solution: 

```scala
    def attempt(triesLeft: Int, errors: List[A]): EitherT[Future, List[A], B] = {
      f.leftFlatMap(error => {
        if (triesLeft > 0) attempt(triesLeft - 1, error +: errors) else EitherT.fromEither(Left(error +: errors))
      })
    }

    attempt(max, Nil)
```

I've used an additional `errors` argument to the recursive function to act as an accumulator for the errors. 

I definitely argree though - if there's a simpler non-recursive version we should definitely go for it!

Reply:
got it, I did write my suggestion without trying it so expected to miss something!
I'll have a quick play with your branch though as I'm intrigued now!  But I don't want to slow things down so happy to go with the recursive one.

Reply:
No not at all! I'm still a scala newbie so these kind of discussions are incredibly helpful for me!

Reply:
I did have a quick play and I don't think there's a function to do it.  We would need a Future aware version of `takeWhile` that returns a `Future[List[A]]`
`Future.sequence` is future aware in the right way, but it would take the whole list, so we would need `Future.sequenceWhile` or `Future.sequenceUntil` which would have to be implemented by us (using fold or recursion) so would not be a great benefit!
Quite interesting to think about it!

Reply:
I've just been having a think about this and it's really hard to implement without recursion!
Why do we want to avoid recursion here? I think Tom's recursive solution is quite neat, and a natural way to reason about this particular problem.

Reply:
yeah I agree there's no well known abstraction for sequenceWhile, and I agree it is fine, I just think that recursion is very low level, and inductive reasoning isn't totally natural if you haven't used it a lot (same goes for folds).

Since all we are trying to describe is "retry a future 3 times/until it succeeds" the recursion is an implementation detail.  We could implement anything with a fold but we would tend to use map/filter/etc because they are simpler abstractions.

In this case though, I do think the implementation is perfectly clear beyond the recursion!

Reply:
Not going to affect this PR, but since we're on the topic, I had a quick google for "traverseWhile" and apparently it's really easy in fs2 (which is a whole other topic that I should learn one day!)
https://gitter.im/typelevel/cats?at=5aac3041c3c5f8b90d7e57ac

Reply:
@johnduffell that fs2 implementation looks nice!

I just had an attempt at a non-recursive version but it's horrible:
```
def apply[A,B](max: Int)(f: => EitherT[Future,A,B])(implicit ec: ExecutionContext): EitherT[Future, List[A], B] =
    // iterates through to max, but only calls f until first success
    (0 to max).foldLeft[EitherT[Future,List[A],B]](EitherT.fromEither(Left(Nil))) { case (acc, _) =>
      acc.leftFlatMap(errors => {
        f.biflatMap[List[A], B](
          err => EitherT.fromEither(Left(err +: errors)),
          value => EitherT.fromEither(Right(value))
        )
      })
    }
```

Reply:
haha nice work Tom! well in my mind fold is recursion using function parameters rather than in primitives, but I'm still impressed.  Also terminating folds early is always a problem in scala, because even with a LazyList, fold is strict in its arguments.
I'm really tempated to implement traverseUntil now, but it's turning into scala wars on @TomPretty 's PR! 😆